cmd/snap/inhibit.go

```diff
@@ -0,0 +1,159 @@
+	// and then either unlocks it or changes to HintInhibitedForRefresh (see
+	// gateAutoRefreshHookHandler in hooks.go).
+	// waitInhibitUnlock will return also on HintNotInhibited.
+	notInhibited, err := waitInhibitUnlock(snapName, runinhibit.HintInhibitedForRefresh, nil)
+	if err != nil {
+		return err
+	}
+	if notInhibited {
+		return nil
+	}
+
+	if isGraphicalSession() && hasZenityExecutable() {
+		return zenityFlow(snapName, hint)
+	}
+	// terminal and headless
+	return textFlow(snapName, hint)
+}
+
+func inhibitMessage(snapName string, hint runinhibit.Hint) string {
+	switch hint {
+	case runinhibit.HintInhibitedForRefresh:
+		return fmt.Sprintf(i18n.G("snap package %q is being refreshed, please wait"), snapName)
+	default:
+		return fmt.Sprintf(i18n.G("snap package cannot be used now: %s"), string(hint))
+	}
+}
+
+var isGraphicalSession = func() bool {
+	return os.Getenv("DISPLAY") != "" || os.Getenv("WAYLAND_DISPLAY") != ""
+}
+
+var hasZenityExecutable = func() bool {
+	return osutil.ExecutableExists("zenity")
+}
+
+func zenityFlow(snapName string, hint runinhibit.Hint) error {
+	zenityTitle := i18n.G("Snap package waiting for update")
+
+	// Run zenity with a progress bar.
+	// TODO: while we are waiting ask snapd for progress updates and send those
+	// to zenity via stdin.
+	zenityDied := make(chan error, 1)
+
+	// TODO: use a dbus API to allow integration with native desktop environment.
+	cmd := exec.Command(
+		"zenity",
+		// [generic options]
+		"--title="+zenityTitle,
+		// [progress options]
+		"--progress",
+		"--text="+inhibitMessage(snapName, hint),
+		"--pulsate",
+	)
+	if err := cmd.Start(); err != nil {
+		return err
+	}
+	// Make sure that zenity is eventually terminated.
+	defer cmd.Process.Signal(os.Interrupt)
+	// Wait for zenity to terminate and store the error code.
+	// The way we invoke zenity --progress makes it wait forever.
+	// so it will typically be an external operation.
+	go func() {
+		zenityErr := cmd.Wait()
+		if zenityErr != nil {
+			zenityErr = fmt.Errorf("zenity error: %s\n", zenityErr)
+		}
+		zenityDied <- zenityErr
+	}()
+
+	if _, err := waitInhibitUnlock(snapName, runinhibit.HintNotInhibited, zenityDied); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func textFlow(snapName string, hint runinhibit.Hint) error {
+	fmt.Fprintf(Stdout, "%s\n", inhibitMessage(snapName, hint))
+	pb := progress.MakeProgressBar()
+	pb.Spin(i18n.G("please wait..."))
+	_, err := waitInhibitUnlock(snapName, runinhibit.HintNotInhibited, nil)
+	pb.Finished()
+	return err
+}
+
+var isLocked = runinhibit.IsLocked
+
+var waitInhibitUnlock = func(snapName string, waitFor runinhibit.Hint, errCh <-chan error) (notInhibited bool, err error) {
+	// Every 0.5s check if the inhibition file is still present.
+	ticker := time.NewTicker(500 * time.Millisecond)
+	defer ticker.Stop()
+loop:
+	for {
+		select {
+		case err := <-errCh:
+			if err != nil {
+				fmt.Fprintf(Stderr, "%s", err)
+			}
+			break loop
```

Comment:
basically from my other comment I think this should read:
```suggestion
			if err != nil {
				return false, err
			}
			break loop
```

Reply:
Agreed, it's more readable (though it should be false, nil). Applied, thanks.